File: problems/901-Online-Stock-Span/solution.go
type StockSpanner struct {
    Prices []int
    Spans []int
}


func Constructor() StockSpanner {
    return StockSpanner{
        Prices: make([]int, 0),
        Spans: make([]int, 0),
    }
}


func (this *StockSpanner) Next(price int) int {
    this.Prices = append(this.Prices, price)

    if len(this.Prices) == 1 {
        this.Spans = append(this.Spans, 1)
    } else {
        curr, prev := len(this.Prices)-1, len(this.Prices)-2
        for {
            if this.Prices[prev] <= this.Prices[curr] {
                prev -= this.Spans[prev]
            } else {
                break
            }

            if prev < 0 {
                break
            }
        }

        this.Spans = append(this.Spans, curr-prev)
    }

    return this.Spans[len(this.Spans)-1]
}


/**
 * Your StockSpanner object will be instantiated and called as such:
 * obj := Constructor();
 * param_1 := obj.Next(price);
 */

File: problems/901-Online-Stock-Span/solution.java
class StockSpanner {
    private List<Integer> days;
    private List<Integer> prices;

    public StockSpanner() {
        days = new ArrayList<>();
        days.add(0);
        prices = new ArrayList<>();
        prices.add(Integer.MAX_VALUE);
    }
    
    public int next(int price) {
        while (prices.get(days.get(days.size() - 1)) <= price) {
            days.remove(days.size() - 1);
        }

        int result = prices.size() - days.get(days.size()-1);

        days.add(prices.size());
        prices.add(price);

        return result;
    }
}

/**
 * Your StockSpanner object will be instantiated and called as such:
 * StockSpanner obj = new StockSpanner();
 * int param_1 = obj.next(price);
 */

